### PR TITLE
Add a unit test for convergence of acquisition function in `GPSampler`

### PR DESCRIPTION
**Pull request by nabenabe0928**
## Motivation

This PR addresses the comment by @toshihikoyanase in the following PR:
- https://github.com/optuna/optuna/pull/5359

## Description of the changes

I added a warning message for the convergence of the acquisition function and its unit test.

### CODE REVIEW

File: optuna/_gp/optim_mixed.py [hunk 1]
```diff
@@ -298,10 +298,13 @@ def optimize_acqf_mixed(
...
     probs[max_i] = 0.0  # We already picked the best param, so remove it from roulette.
     probs /= probs.sum()
+    n_non_zero_probs_improvement = np.count_nonzero(probs > 0.0)
     # n_additional_warmstart becomes smaller when study starts to converge.
     n_additional_warmstart = min(
-        n_local_search - len(warmstart_normalized_params_array) - 1, np.count_nonzero(probs > 0.0)
+        n_local_search - len(warmstart_normalized_params_array) - 1, n_non_zero_probs_improvement
     )
+    if n_additional_warmstart == n_non_zero_probs_improvement:
+        _logger.warning("Study already converged, so we reduce the number of local search.")
```
Comment (contramundum53) [inline optuna/_gp/optim_mixed.py:307]:
```suggestion
        _logger.warning("Study already converged, so the number of local search is reduced.")
```

File: tests/samplers_tests/test_gp.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    ]
+    optimal_trials = [
+        optuna.create_trial(value=0.0, params={"x": 0.0}, distributions=dists) for _ in range(10)
+    ]
+    # A large `optimal_trials` causes the instability in the kernel inversion, leading to
+    # instability in the variance calculation.
+    study.add_trials(uniform_trials + uniform_near_optimal_trials + optimal_trials)
+    study.optimize(lambda t: t.suggest_float("x", 0.0, 1.0), n_trials=1)
+    msg = "Study already converged, so we reduce the number of local search"
+    assert msg in caplog.text, "To the PR author, did you change the kernel implementation?"
```
Comment (contramundum53) [inline tests/samplers_tests/test_gp.py:32]: I don't know if checking the specific warning message is a good practice. Also, this test is fragile to change in any implementation details of GP.
How about mocking `eval_acqf` instead?

File: optuna/_gp/optim_mixed.py [hunk 3]
```diff
@@ -262,7 +262,7 @@ def local_search_mixed(
             # Parameters not changed from the beginning.
             return (best_normalized_params, best_fval)
 
-    _logger.warn("local_search_mixed: Local search did not converge.")
+    _logger.warning("local_search_mixed: Local search did not converge.")
```
Comment (eukaryo) [inline optuna/_gp/optim_mixed.py:265]: Could you make this line covered by tests?

File: tests/samplers_tests/test_gp.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+from __future__ import annotations
+
+from _pytest.logging import LogCaptureFixture
+
+import optuna
+
+
+def test_after_convergence(caplog: LogCaptureFixture) -> None:
```
Comment (HideakiImamura) [inline tests/samplers_tests/test_gp.py:8]: It is overkill to create a study and calling `study.optimize` for the tests of the internal `_gp` module. To limit the range of impact, unit tests should be written with the minimum necessary interfaces. How about just creating the `np.ndarray` and give it to `local_search_mixed`?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I have several comments. PTAL.

**Comment by nabenabe0928** [comment]
Review ready!

**Comment by nabenabe0928** [comment]
@HideakiImamura @eukaryo This PR is review ready!

File: tests/samplers_tests/test_gp.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    acqf_params = acqf.create_acqf_params(
+        acqf_type=acqf.AcquisitionFunctionType.LOG_EI,
+        kernel_params=kernel_params,
+        search_space=search_space,
+        X=X[:, np.newaxis],
+        Y=score_vals,
+    )
+    caplog.clear()
+    optuna.logging.enable_propagation()
+    assert len(caplog.text) == 0, "Cap log must be cleared before the optimization."
```
Comment (HideakiImamura) [inline tests/samplers_tests/test_gp.py:43]:
```suggestion
    This assertion does not test anything, so can be removed.
```

File: tests/samplers_tests/test_gp.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    optim_mixed.optimize_acqf_mixed(acqf_params, rng=np.random.RandomState(42))
+    # len(caplog.text) > 0 means the optimization has already converged.
+    assert len(caplog.text) > 0, "To the PR author, did you change the kernel implementation?"
```
Comment (HideakiImamura) [inline tests/samplers_tests/test_gp.py:46]:
```suggestion
    assert len(caplog.text) > 0
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I have two minor comments. PTAL.

File: tests/samplers_tests/test_gp.py (hunk 6 above)
Comment (nabenabe0928) [reply tests/samplers_tests/test_gp.py:45]:
```suggestion
    assert len(caplog.text) > 0, "Did you change the kernel implementation?"
```

File: optuna/_gp/optim_mixed.py (hunk 3 above)
Comment (eukaryo) [reply optuna/_gp/optim_mixed.py:265]: Sorry, I must have misunderstood, but this comment is outside the scope of the PR.

**Comment by eukaryo** [comment]
@not522 Could you review this PR?

**Review by not522 — APPROVED** [review]
LGTM

**Pull request merged**